### ISSUE TITLE
Feature request: emoji support

### ISSUE THREAD

**Issue by atoktoto** [mabe02/lanterna#505]
The code `textGraphics.putString(0, 3, "🍕")` results in two question mark characters being displayed in the terminal. At the same time `System.out.println("🍕")` works as intended (at least in terminal emulators supporting emojis: ie. the new Windows Terminal).

Is it possible to support this use-case?
I guess the emoji codepoints do not fit into `char` type that is used in `Terminal.putCharacter` so this would require major changes.

**Comment by rednoah** [comment]
`com.ibm.icu.text.BreakIterator` can be used to iterate grapheme clusters (i.e. single-width user-perceived characters):
```java
import com.ibm.icu.text.BreakIterator;
```
```java
public static List<String> getGraphemeClusters(String self) {
	List<String> characters = new ArrayList<String>(self.length());
	BreakIterator i = BreakIterator.getCharacterInstance();
	i.setText(self);
	for (int begin = 0, end = 0; (end = i.next()) != BreakIterator.DONE; begin = i.current()) {
		characters.add(self.substring(begin, end));
	}
	return characters;
}
```

**Comment by atoktoto** [comment]
Seems certainly doable but would require a significant change (or addition) to the Lanterna interfaces. Currently, `char` and `TextCharacter` seems to be the center of the whole operation. Replacing it with `String` representing a single grapheme cluster seems wasteful (in terms of memory) for the general case and can make it less legible: `void putCharacter(String s)` looks wrong :D

**Comment by mabe02** [comment]
I don't see why emoji wouldn't work with the current system, given that we can do CJK characters just fine. I'll investigate, maybe it's the terminal encoding that needs to be updated.

**Comment by rednoah** [comment]
Here's what I get with `lanterna 3.0.3` for a file name such as `THAIบุฟเฟต์EMOJI👩‍👩‍👦‍👦.txt`:
* Works for Thai
* Does not work for complex Emoji

![Screen Shot 2020-08-30 at 09 54 06](https://user-images.githubusercontent.com/1589068/91650205-eb37e100-eaa6-11ea-88e5-b05aa2d95757.png)

That being said, neither `iTerm` nor `Terminal` render this particular Emoji correctly either:
```
 ls
THAIบุฟเฟต์EMOJI👩?👩?👦?👦.txt
```

**Comment by rednoah** [comment]
EDIT: `บุฟเฟต์` does render correctly, but the layout does not account for compound characters `บุ` and `ต์` taking up only `1` character (even though it's `2 chars`) and so the layout is off by `2` here:
![Screen Shot 2020-08-30 at 10 10 32](https://user-images.githubusercontent.com/1589068/91650382-2a673180-eaa9-11ea-92dc-555ef0c09745.png)

```
$ ls
TEST.mp4
บุฟเฟต์.mp4
```

**Comment by mabe02** [comment]
Interesting, so the CJK detector incorrectly flags `บุ` as two text characters wide?

**Comment by mabe02** [comment]
Ok, I see the problem now. Java char type isn't able to store emoji:
https://developers.redhat.com/blog/2019/08/16/manipulating-emojis-in-java-or-what-is-%F0%9F%90%BB-1/
Slightly unexpected. Will see what we can do about this.

**Comment by rednoah** [comment]
Yes, `บุ` is `2` code points. It even requires hitting `DELETE` twice to delete the entire character. Hitting `DELETE` once only change `บุ` to `บ`. Kinda like [https://en.wikipedia.org/wiki/Unicode_equivalence#Example](NFD).

**Comment by mabe02** [comment]
The problem I'm finding is that even `"บุ".length()` returns 2... 
I'm trying to change the internal representation of TerminalCharacter to String, but it's tricky to know if the character should be considered single- or double-width, given Java provides little guidance. I'd like to avoid hard-coding unicode page references if possible...

**Comment by mabe02** [comment]
Have been browsing articles and it really seems like while we can get the number of code points, there's no way to know if these code points are combined into a single character, or if that character is double or single width!

**Comment by rednoah** [comment]
Yep, pretty much. `lanterna` effectively can't predict how a terminal window is going to render the text, because it depends on the version of unicode used by the text renderer. Though we can generally assume that long-established unicode sequences like `บุ` will work just fine, while recent additions like  `👩‍👩‍👦‍👦` are likely to not work.

**Comment by mabe02** [comment]
Ok, so here's what we'll do. In 3.0 we'll restrict TextCharacter to BMP only, with an override if you really know what you're doing. In 3.1, also restrict but change to use String internally and let you supply your own "String" character for complicated emoji. Will try this out.

**Comment by mabe02** [comment]
Ok, I misunderstood the BMP plane again. I've just blocked 3.0 from creating TextCharacters from surrogate char:s at least. So next will use the BreakIterator above to in 3.1 to try to group characters.

**Comment by mabe02** [comment]
Okay, I've re-worked TextCharacter to support this:
https://github.com/mabe02/lanterna/pull/508

**Comment by mabe02** [comment]
Ok, code is merged. If you clone and build release/3.1 (I'll do another release in a week or so) you should be able to print emoji as double-width and your magic บุ character only occupying one column. Please try it out and report back before I close this.

**Comment by MVoloshin** [comment]
Cant print BOMB character "\uD83D\uDCA3" or 💣 using Lanterna 3.2.0-master. I just get two rectangles(

**Comment by rsv-code** [comment]
@mabe02 I too can't get emojis working. I'm testing on 3.1.2 and 3.2.0-alpha1 from maven. I'm only getting rectangles in the swing window and question marks in the normal terminal window in Ubuntu. Is there something I need to do to enable them or to display them. Thanks for your assistance!

**Comment by avl42** [comment]
Emojis are a bit "complicated" - they are incompatible with lanterna's
current approach of 1 (java) "char" per
screen-position (cell).

The complicated solution would be to store Strings for each cell, but that
would lead to a more or less
complete rewrite of the whole library.

Eventually it might happen, and it will not only enable emojis, but also
enable some glyphs in
some scripts, that get combined out of several characters.  But it may be
too early to start
holding your breath for it.

**Comment by mabe02** [comment]
Works for me on Ubuntu:
![image](https://github.com/user-attachments/assets/aec99e0b-9024-46ab-a802-36fb0cc22bf9)
Maybe it's a font problem? I just did this:
terminal.putString("\uD83D\uDCA3");

(modified SwingTerminalTest.java)

**Comment by mcarleio** [comment]
I just stumbled upon this issue as well, here are my findings:

- `terminal.putString("🏫=House")`:
  -  works in any terminal that relies on `GraphicalTerminalImplementation`  ✅
  - does not work in any `StreamBasedTerminal` 🚫
- `screen.newTextGraphics().putString(0, 0, "🏫=House")`
  - does not work at all (prints only first char of emoji) 🚫
  - for any terminal that relies on `GraphicalTerminalImplementation`, this is a workaround ✅👷 
    ```java
    var tg = screen.newTextGraphics()
    var textCharacters = TextCharacter.fromString(line);
    int prevDoubleCharCount = 0;
    for (int i = 0; i < textCharacters.length; i++) {
      TextCharacter textCharacter = textCharacters[j];
      tg.setCharacter(i + prevDoubleCharCount, 0, textCharacter);
      if (textCharacter.isDoubleWidth()) {
          prevDoubleCharCount++;
      }
    }
    ```
- `new Label("🏫=House")`
  - does not work at all (prints only first char of emoji) 🚫
  - for any terminal that relies on `GraphicalTerminalImplementation`, there is a workaround: you have to define a custom `ComponentRenderer<Label>` which uses the previous workaround ✅👷


Would be lovely if someone could have a look on this and fix the `StreamBasedTerminal` implementation and the `TextGraphics` implementation.

**Comment by mcarleio** [comment]
The main issue for the `StreamBasedTerminal.putString` method seems to be the processing of every character on its own.
If you replace the implementation with `writeToTerminal(string.getBytes(terminalCharset));` it does print the emojis on the `UnixTerminal`.

This is a hacky workaround and it does not respect the `Symbols` mapping (i.e. does not call `convertToVT100`). Not sure what that might break or change.

**Comment by avl42** [comment]
Just some information about the `convertToVT100`:
some terminals that do not yet support utf-8 (or are configured not to
support it)
These terminals still offer an "alternative charset" that contains e.g.
linedrawing
characters.

The said function allows boxes(as in window frames) to be drawn even
without utf8.

In a nutshell:  if you want to display emojis to a terminal, then if the
terminal
needed the conversion done in that method, then it certainly wouldn't
display
the emoji anyway.

Now it seems that there are ways to "cheat" a sequence of characters to the
terminal, but that is not working at all with screen or textgui.

One idea could be to reserve one of the possible "char" values, add a String
to the TextCharacter class, and if that particular value is in the char
field,
then the String would be written out.   This paragraph describes an idea
on how to add emojis to lanterna without breaking compatibility completely.

This proposed change would involve changing most methods in Graphics
and Screen and TextGui to recognize multi-char sequences and produce
TextCharacter instances with the character set to the reserved value, and
the string set to the full emoji.    On output the special value of
TextCharacter's
character field would be recognized and the String written out instead.

**Comment by mcarleio** [comment]
That sounds like a lot of work, but probably would be a good long term solution.

For short term solution, I could create a PR which just changes the implementation of 
- `StreamBasedTerminal.putString`
- `AbstractTextGraphics.putString`
- `Label.createDefaultRenderer`
